src/System.Management.Automation/utils/Telemetry.cs

```diff
@@ -0,0 +1,407 @@
+
+        // the telemetry failure string
+        private const string _telemetryFailure = "TELEMETRY_FAILURE";
+
+        // Telemetry client to be reused when we start sending more telemetry
+        private static TelemetryClient s_telemetryClient { get; set; }
+
+        // Set this to true to reduce the latency of sending the telemetry
+        private static bool s_developerMode { get; set; }
+
+        // the unique identifier for the user, when we start we
+        private static Guid s_uniqueUserIdentifier { get; set; }
+
+        // the session identifier
+        private static Guid s_sessionId { get; set; }
+
+        /// <summary>
+        /// We send telemetry only a known set of modules.
+        /// If it's not in this list, then we report anonymous.
+        /// </summary>
+        private static string[] s_knownModuleNames = new[]
+        {
+                "Microsoft.PowerShell.Archive",
+                "Microsoft.PowerShell.Host",
+                "Microsoft.PowerShell.Management",
+                "Microsoft.PowerShell.Security",
+                "Microsoft.PowerShell.Utility",
+                "PackageManagement",
+                "Pester",
+                "PowerShellGet",
+                "PSDesiredStateConfiguration",
+                "PSReadLine",
+                "ThreadJob",
+        };
+
+        // use a hashset for quicker lookups.
+        private static HashSet<string> s_knownModules;
+
+        /// <summary>Gets a value indicating whether telemetry can be sent.</summary>
+        public static bool CanSendTelemetry { get; private set; }
+
+        /// <summary>
+        /// Initializes static members of the <see cref="ApplicationInsightsTelemetry"/> class.
+        /// Static constructor determines whether telemetry is to be sent, and then
+        /// sets the telemetry key and set the telemetry delivery mode.
+        /// Creates the session ID and initializes the HashSet of known module names.
+        /// Gets or constructs the unique identifier.
+        /// </summary>
+        static ApplicationInsightsTelemetry()
+        {
+            // If we can't send telemetry, there's no reason to do any of this
+            CanSendTelemetry = !GetEnvironmentVariableAsBool(name: _telemetryOptoutEnvVar, defaultValue: false);
+            if (CanSendTelemetry)
+            {
+                s_telemetryClient = new TelemetryClient();
+                TelemetryConfiguration.Active.InstrumentationKey = _psCoreTelemetryKey;
+                TelemetryConfiguration.Active.TelemetryChannel.DeveloperMode = s_developerMode;
+                s_sessionId = Guid.NewGuid();
+
+                // use a hashset when looking for module names, it should be quicker than a string comparison
+                s_knownModules = new HashSet<string>(s_knownModuleNames, StringComparer.OrdinalIgnoreCase);
+                s_uniqueUserIdentifier = GetUniqueIdentifier();
+                s_developerMode = false;
```

Comment:
Do we need this static field? It's not used anywhere else, so maybe just remove it and change to `TelemetryConfiguration.Active.TelemetryChannel.DeveloperMode = false`?

Reply:
fixed